xtask/src/set_dep_version.rs

```diff
@@ -0,0 +1,273 @@
+            match &args.spec.version {
+                Some(version) => Value::from(version.to_string()),
+                None => Value::from("*"),
+            },
+        )]);
+
+        copy_dependency_features(&mut new_dep, dep);
+
+        *dep = new_dep.into();
+        simplify_dependency_table(dep)
+    }
+
+    deps.fmt();
+    deps.sort_values();
+
+    eprintln!("[workspace.dependencies]");
+    for (key, dep) in deps.iter().filter(|(key, _)| args.dep.owns(key)) {
+        eprintln!("{key} = {dep}");
+    }
+}
+
+fn edit_patch(cargo_toml: &mut DocumentMut, args: &Args) {
+    let patch = cargo_toml["patch"].as_table_mut().unwrap()["crates-io"]
+        .as_table_mut()
+        .unwrap();
+
+    // Clear any existing entries for this dependency.
+    for crate_name in args.dep.crates() {
+        patch.remove(crate_name);
+    }
+
+    // Leave this section as-if if we are requested to just use a specific version.
+    if args.spec.rev.is_some() || args.spec.branch.is_some() || args.spec.path.is_some() {
+        // Patch all Cairo crates that exist, even if this project does not directly depend on them,
+        // to avoid any duplicates in transient dependencies.
+        for &dep_name in args.dep.crates() {
+            let mut dep = InlineTable::new();
+
+            // Add a Git branch or revision reference if requested.
+            if args.spec.rev.is_some() || args.spec.branch.is_some() {
+                dep.insert("git", args.dep.repo().into());
+            }
+
+            if let Some(branch) = &args.spec.branch {
+                dep.insert("branch", branch.as_str().into());
+            }
+
+            if let Some(rev) = &args.spec.rev {
+                dep.insert("rev", rev.as_str().into());
+            }
+
+            // Add local path reference if requested.
+            // For local path sources, Cargo is not looking for crates recursively therefore, we
+            // need to manually provide full paths to Cairo workspace member crates.
+            if let Some(path) = &args.spec.path {
+                dep.insert(
+                    "path",
+                    path.join("crates")
+                        .join(dep_name)
+                        .to_string_lossy()
+                        .into_owned()
+                        .into(),
+                );
+            }
+
+            patch.insert(dep_name, dep.into());
+        }
+    }
+
+    patch.fmt();
+    patch.sort_values();
+
+    eprintln!("[patch.crates-io]");
+    for (key, dep) in patch.iter() {
+        eprintln!("{key} = {dep}");
+    }
+}
+
+impl DepName {
+    fn crates(&self) -> &'static [&'static str] {
+        match self {
+            DepName::Cairo => {
+                // List of library crates published from the starkware-libs/cairo repository.
+                // One can get this list from the `scripts/release_crates.sh` script in that repo.
+                // Keep this list sorted for better commit diffs.
+                &[
```

Comment:
Yes, but the whitelist would be easier to maintain, right? 🤔

Reply:
you meant a blacklist, right?

I don't know - cannot predict what crates will appear more frequently in the future, because this is the source of maintenance burden

Reply:
Yes, blacklist :p 

Apart from frequency, I feel like a new crate that should be whitelisted can be harder to spot, apart from manually checking the Cairo repo, but it can also easily become our dep - don't you think? 🤔

Reply:
I tried implementing this, but complexity explodes:
1. I still need whitelists for other tools (cairols, future cairo-lint etc.)
2. I cannot trust the lock file at the time of starting this script because at the moment of upgrading it might not have a new crate yet. I would have to make an upgrade, do patches, and repeat this process once more with an updated list of crates.

One alternative idea I got is to `curl -L https://raw.githubusercontent.com/starkware-libs/cairo/refs/heads/main/scripts/release_crates.sh` and try to extract these crates from the response, but I doubt we want to go that way. wdyt?

Reply:
What is the potential danger if the compiler teams introduces a new crate and we forget about it here? Duplicated deps in `Cargo.lock` again + incompabilities with LS/cairo-lint?

If they introduce a dep that we don't use I guess we don't really care, since it will be resolved to latest available version (thus the correct one) for both nightlies (git versions) and full releases (crates.io), right?
So we only really care if we forget about new deps that we start using. Maybe the `key.get().starts_with("cairo-lang-"))` logic was fine all along?

Reply:
> One alternative idea I got is to curl -L https://raw.githubusercontent.com/starkware-libs/cairo/refs/heads/main/scripts/release_crates.sh and try to extract these crates from the response, but I doubt we want to go that way

Sounds like this could be a good warning to show and expect some human reaction? 🤔

>  What is the potential danger if the compiler teams introduces a new crate and we forget about it here? Duplicated deps in Cargo.lock again + incompabilities with LS/cairo-lint?

I believe we need to patch all dependencies used by the Cairo-LS / Cairo-lint, otherwise the versions defined in LS/Lint Cargo.toml will be used as dependencies of LS/lint.

Reply:
>I believe we need to patch all dependencies used by the Cairo-LS / Cairo-lint, otherwise the versions defined in LS/Lint Cargo.toml will be used as dependencies of LS/lint.

They will be used, but if the deps are synced between LS and lint, then it wouldn't be a problem if we don't include this dep in Scarb's Cargo.toml, right?

Reply:
I suggest the following: let's merge this PR with this list hardcoded and I will continue experimenting on this with the `curl` approach as I believe it might solve the issue